cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/ConfiguratorTask.java

```diff
@@ -0,0 +1,135 @@
+import io.cdap.cdap.common.id.Id;
+import io.cdap.cdap.common.internal.worker.RunnableTask;
+import io.cdap.cdap.common.internal.worker.RunnableTaskContext;
+import io.cdap.cdap.common.io.Locations;
+import io.cdap.cdap.config.CConfigurationCodec;
+import io.cdap.cdap.internal.app.ApplicationSpecificationAdapter;
+import io.cdap.cdap.internal.app.deploy.InMemoryConfigurator;
+import io.cdap.cdap.internal.app.deploy.pipeline.ConfiguratorConfig;
+import io.cdap.cdap.internal.app.runtime.artifact.ArtifactRepository;
+import io.cdap.cdap.internal.app.runtime.artifact.PluginFinder;
+import io.cdap.cdap.security.impersonation.EntityImpersonator;
+import io.cdap.cdap.security.impersonation.Impersonator;
+import org.apache.twill.filesystem.Location;
+import org.apache.twill.filesystem.LocationFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * ConfiguratorTask is a RunnableTask for performing the configurator config.
+ */
+public class ConfiguratorTask implements RunnableTask {
+  private static final Gson GSON = ApplicationSpecificationAdapter.addTypeAdapters(new GsonBuilder())
+    .registerTypeAdapter(CConfiguration.class, new CConfigurationCodec()).create();
+  private static final Logger LOG = LoggerFactory.getLogger(ConfiguratorTask.class);
+
+  private final CConfiguration cConf;
+
+  @Inject
+  public ConfiguratorTask(CConfiguration cConf) {
+    this.cConf = cConf;
+  }
+
+  @Override
+  public void run(RunnableTaskContext context) throws Exception {
+    Injector injector = Guice.createInjector(new ConfiguratorTaskModule(cConf));
+    Configurator configurator = injector.getInstance(Configurator.class);
+    configurator.run(context);
+  }
+
+  /**
+   * Configurator performs the configurator config.
+   */
+  public static class Configurator {
+    private final Impersonator impersonator;
+    private final PluginFinder pluginFinder;
+    private final ArtifactRepository artifactRepository;
+    private final LocationFactory locationFactory;
+
+    @Inject
+    public Configurator(Impersonator impersonator, PluginFinder pluginFinder,
+                        ArtifactRepository artifactRepository, LocationFactory locationFactory) {
+      this.impersonator = impersonator;
+      this.pluginFinder = pluginFinder;
+      this.artifactRepository = artifactRepository;
+      this.locationFactory = locationFactory;
+    }
+
+    public void run(RunnableTaskContext context) throws Exception {
+      ConfiguratorConfig config = GSON.fromJson(context.getParam(), ConfiguratorConfig.class);
+
+      // Getting the pipeline app from appfabric
+      LOG.info(String.format("Fetching artifact '%s' from app-fabric to create artifact class loader.",
+                             config.getArtifactId().getArtifact()));
+      Location artifactLocation = Locations.getLocationFromAbsolutePath(
```

Comment:
I don't understand this logic. Can you explain where are we writing the artifact to? I feel like we need better encapsulation related to artifact fetching, expansion, and caching.

Reply:
As of right now, we're writing the artifact to the same path that it lives in GCS. This path doesnt live on the PD so there is no caching between restarts but if we dont restart the pod we have caching.

I agree this needs to be encapsulated better, that will be done with the file fetcher sidecar work. Then this will be refactored to rely on that.